### DISCUSSION TITLE
Object sorting on ByteArray.writeObject()

### DISCUSSION THREAD

**Discussion by htmiel** [airsdk/Adobe-Runtime-Support#923] [Ideas]
Speaking of sorting JSON/Object in AS3, would it be possible to add such a sorting to the `ByteArray.writeObject()` function?

Currently, whenever I write the same object to a byte array, the output byte array is always different (has a different hash), probably because the same properties are written in a random order.

It would be nice if the `ByteArray.writeObject()` output would be identical when writing an Object containing the exact same information.

**Comment by ajwfrost** [comment]
Do you mean, if you just call `byteArray1.writeObject(obj)` and `byteArray2.writeObject(obj)` (i.e. with the same object written into two separate byte arrays), that you end up with different byte arrays? Interesting....
Part of me wonders whether this is deliberate, done as part of the security hardening that Adobe did a few years back. But we can check :-)

**Reply by htmiel** [reply]
Sorry, I just tested again, and it seems 'ByteArray.writeObject()' does indeed output the same bytes when writing the same AS3 object.

I have a tool to convert an XLSX to AS3/AMF bytes file, which produces different outputs even if I don't change the XLSX input file; I'll look for another source of this randomness...

Again, I apologize for the false alarm.

**Reply by ajwfrost** [reply]
No worries, it's all quite interesting :-)    I had wondered about randomness coming in if there's any timestamps being added to a stream somewhere?

The AMF conversion is a little different to the ByteArray.writeObject method - basically, the 'encoding' provided will switch how it works.

Good luck!

**Reply by htmiel** [reply]
I managed to reproduce the issue.

Running the code below will create the same Object 5 times, and write it to 5 files. The 5 files will be identical (hash).
Running the same code again, will create 5 more files, identical between each other, but different from the first 5 files...

Looks like writing the same object in the same "session", gives the same output, but running the app again, it's different? I tested with AIR 33.1.1.476 and built a Windows captive app.

```
for(i=0; i<5; i++)
{
	objTest = {
		processes: {
			"a1": {
				param1:"1",
				param2:0,
				property:"a"
			},
			"a2": {
				param1:"2",
				param2:0,
				property:"b"
			},
			"a3": {
				param1:"3",
				param2:0,
				property:"c"
			}
		}
	}

	var dataBytes:ByteArray = new ByteArray()
	dataBytes.writeObject(objTest)

	var fileStream:FileStream = new FileStream()
	fileStream.open(File.desktopDirectory.resolvePath(i + ".obj"), FileMode.WRITE)
	fileStream.writeBytes(dataBytes, 0, dataBytes.length)
	fileStream.close()
}
```

The same can be seen when saving an object to file with `ByteArray.writeObject()`, and then (in a separate session) reading the object back with `ByteArray.readObject()` and writing it again to a file with `ByteArray.writeObject()`. The source and final files will have different hashes, even though they contain the same information.

**Reply by ajwfrost** [reply]
Yes, looking at this, when you run it at different times, you get the object properties being output in a different order. Curious! we can look at it if it would be useful for it to be consistent (but will double-check on the security implications...)

**Reply by htmiel** [reply]
Thank you for taking a look at it.

This is not a critical problem; the inconvenience is that I have an automated update system, based on file hashes, and updates get triggered when an AS3 object is re-written to file even if the information inside did not change.

It would be nice to have consistent output, but it's not critical.

**Reply by ajwfrost** [reply]
Having looked at how properties are stored... they're put into a hash table and the ordering is just based on the memory addresses of the elements being stored. So no real way to make it consistent between different runs without destroying the efficiency.

What we could perhaps do is create a new ActionScript method that instructs the object to re-order its dynamic properties so that they're stored alphabetically (though that would then be changed whenever the hash table grew) - or to create an extra setting for "writeObject" so that we don't just iterate through the dynamic properties but instead list them and sort them somehow first...

**Reply by htmiel** [reply]
A new AS3 method or an extra optional parameter to `ByteArray.writeObject(object:*):void`, whichever would be the minimally invasive method is fine. No problem if I would need to call a "sorting" function on an Object before writing it to a byte array.

I definitely would not want anything else to be affected, like general performance or stability of the Object.

**Reply by ajwfrost** [reply]
If you sorted an object, but it contained other objects, then that might still be an issue.. so perhaps it's easier to update the writeObject function so that for the duration of that function, we re-order the object's dynamic properties. This would then mean things are consistent between different instances of the app, even if the order isn't necessarily fully alphabetical (e.g. if you have a mixture of defined properties plus dynamic ones.. hopefully a fairly rare case though!).

But you know that if you defined the objects and used an array then it should be consistent e.g.
```
class ProcessInfo {
  public var param1 : String;
  public var param2 : uint;
  public var property: String;
  public ProcessInfo(p1 : String, p2 : uint, prop : String) {
    param1 = p1; param2 = p2; property = prop;
  }
}
class Process {
  public var name : String;
  public var info : ProcessInfo;
  public Process(_name : String, _info : ProcessInfo) {
    name = _name; info = _info;
  }
}

objTest = {
  processes: [
    new Process("a1", new ProcessInfo("1", 0, "a")),
    new Process("a2", new ProcessInfo("2", 0, "b")),
    new Process("a3", new ProcessInfo("3", 0, "c"))
  ]
};
```
Basically -> if you define the class traits ahead of time, then these are embedded into the ActionScript byte code definitions within your SWF, and will be used a bit more efficiently as "slots". If you use a dynamic object and just add previously-undefined properties to an object, then it uses the hash table mechanism for run-time definitions - which means any property look-up is a little slower (i.e. the compiler doesn't know about it so it's a run-time lookup done via iteration, rather than the compiler knowing what slot you are referencing).

Anyway. That was a bit of an aside :-)  We can have a play here to see if adding a new 'sort' option would be straightforward, I can't see it being much of a change..

thanks

**Reply by htmiel** [reply]
Sorry, I forgot to mention that my use case is only with dynamic objects, which I write to files to store arbitrary information (without a predefined structure like in a class), like other dynamic objects, arrays (of dynamic objects), strings, byte arrays (which I cannot save to JSON), etc.

**Reply by ajwfrost** [reply]
Okay so here's an interesting find! If you look at the `ObjectEncoding` class, there's a static property that you can set which is by default null - and null means that the existing C++ code is used to iterate through the dynamic properties and output them in the order in which they're found in the hash table:
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/net/ObjectEncoding.html#dynamicPropertyWriter

But if you create an object and implement the `IDynamicPropertyWriter` interface, then the C++ code will instead just call your object's `writeDynamicProperties` method.
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/net/IDynamicPropertyWriter.html

You get given the object being serialized, plus an `IDynamicPropertyOutput` reference, where you just call `writeDynamicProperty`.
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/net/IDynamicPropertyOutput.html

So: basically to achieve what you want here, you can create a utility object that's then used as a callback for each object you're serializing, and this can do a "for (var key:String in obj)" to put all the property names into an array, call Array.sort(), and then output the property names/values in alphabetical order.

TBH makes me think there should be an implementation of this included in AIR already, so you could just do:
`ObjectEncoding.dynamicPropertyWriter = ObjectEncoding.sortedPropertyWriter;`
and we could then implement this `sortedPropertyWriter` in C++ to make it a lot more efficient...

**Reply by htmiel** [reply]
Thank you for the detailed reply Andrew. I did not know about the method you described, and it does exactly what I needed! I tested it with the sample code I posted earlier, and the output file hash is now always the same when running it multiple times.

There's no problem if this method is slower than a C++ implementation, since I would use this only for internal tools. I'll now add this to all my custom file formats for consistent results.

Your help is much appreciated once again!
